From 83948d6bcfca41893c1d3b69763026629857cccf Mon Sep 17 00:00:00 2001
From: Zeeshan Ali (Khattak) <zeeshanak@gnome.org>
Date: Fri, 13 Nov 2009 20:21:04 +0200
Subject: [PATCH] Handle 'NOTIFY' received before subscription completes

There is sometimes a race-condition in which the notification events are
are received from the server *before* the subscribtion is complete. This
patches fixes handling of that case.
---
 libgupnp/gupnp-service-proxy.c |   18 +++++++++++++-----
 1 files changed, 13 insertions(+), 5 deletions(-)

diff --git a/libgupnp/gupnp-service-proxy.c b/libgupnp/gupnp-service-proxy.c
index 19460b6..9098ce1 100644
--- a/libgupnp/gupnp-service-proxy.c
+++ b/libgupnp/gupnp-service-proxy.c
@@ -1532,16 +1532,24 @@ server_handler (SoupServer        *soup_server,
         }
 
         hdr = soup_message_headers_get (msg->request_headers, "SID");
-        if (hdr == NULL ||
-            (proxy->priv->sid == NULL ||
-             strcmp (hdr, proxy->priv->sid) != 0)) {
-                /* No SID or not ours */
+        if (G_UNLIKELY (hdr == NULL)) {
+                /* No SID */
                 soup_message_set_status (msg, SOUP_STATUS_PRECONDITION_FAILED);
 
                 return;
         }
 
-        handle_notify (proxy, msg);
+        if (proxy->priv->sid == NULL) {
+                if (proxy->priv->subscribed) {
+                        /* Subscription in progress */
+                        handle_notify (proxy, msg);
+                }
+        } else if (strcmp (hdr, proxy->priv->sid) != 0) {
+                /* Not ours */
+                soup_message_set_status (msg, SOUP_STATUS_PRECONDITION_FAILED);
+        } else {
+                handle_notify (proxy, msg);
+        }
 }
 
 /*
-- 
1.6.1

